feat(#1226): Check `eo-maven-plugin` version during caching

volodya-lombrozo opened the pull request:
Added `CacheVersion` class that encapsulates the logic of choosing whether `eo-maven-plugin` version is cacheable or not.

Closes: #1226

volodya-lombrozo commented:
@mximp What do you think if we replace `Footprints` [implementation](https://github.com/objectionary/eo/pull/1276) with the new `CacheVersion` class instead?

mximp commented on eo-maven-plugin/src/main/java/org/eolang/maven/footprint/CacheVersion.java:
```diff
@@ -0,0 +1,82 @@
+import java.util.Objects;
+
+/**
+ * Cache version.
```
@volodya-lombrozo Can you please elaborate the description. It's not quite clear what cache is meant. Maybe add what format is represented?

mximp commented on eo-maven-plugin/src/main/java/org/eolang/maven/footprint/CacheVersion.java:
```diff
@@ -0,0 +1,82 @@
+    private static final String[] NOT_CACHEABLE = {"0.0.0", "SNAPSHOT"};
+
+    /**
+     * Version of the eo-maven-plugin.
```
@volodya-lombrozo how is this related to `eo-maven-plugin`? Is some format implied?

mximp commented on eo-maven-plugin/src/main/java/org/eolang/maven/footprint/CacheVersion.java:
```diff
@@ -0,0 +1,82 @@
+    private final String version;
+
+    /**
+     * Hash of the object.
```
@volodya-lombrozo is this for single object?

mximp commented on eo-maven-plugin/src/main/java/org/eolang/maven/footprint/CacheVersion.java:
```diff
@@ -0,0 +1,82 @@
+     * Get cache path.
+     * @return Path.
+     */
+    Path cache() {
```
@volodya-lombrozo I believe better name would be `path()`. `version.path()` seems to be more natural.

volodya-lombrozo commented:
@mximp I've renamed `cache()` -> `path()` and added more informative description. Could you check again, please? Maybe we can improve the description even more?

mximp commented:
@volodya-lombrozo I still believe that referring to `eo-maven-plugin` is misleading. Can't this class be used for any other version? I see the description might be as follows:
```java
/**
  * A version consisting of two parts: 1) base which is maven-style version string
  * and 2) hash string containing git hash. The two parts define the cacheable location be method {@link #path}
  * If base is empty or contains '0.0.0' or 'SNAPSHOT' the version is considered non-cacheable.
*/
```

volodya-lombrozo commented:
> @volodya-lombrozo I still believe that referring to `eo-maven-plugin` is misleading. Can't this class be used for any other version? I see the description might be as follows:
> 
> ```java
> /**
>   * A version consisting of two parts: 1) base, which is maven-style version string
>   * and 2) hash string, containing git hash. The two parts define the cacheable location by method {@link #path}
>   * If base is empty or contains '0.0.0' or 'SNAPSHOT' the version is considered non-cacheable.
> */
> ```

@mximp Thanks! Updated. Could you have a look, again, please?

volodya-lombrozo commented:
@yegor256 Could you merge that changes, please?

yegor256 commented:
@rultor merge

rultor commented:
> @rultor merge

@yegor256 OK, I'll try to merge now. You can check the progress of the merge [here](https://www.rultor.com/t/30455-1448054087)

rultor commented:
> @rultor merge

@yegor256 Done! FYI, the full log is [here](https://www.rultor.com/t/30455-1448054087) (took me 7min)
